### PR TITLE
Core/Unit: Fix removal of auras on evade issues

### LINKED ISSUE #14347: [3.3.5][6.x] Guardians evade issue
rev. 68c4ea314443

After a combat or in_combat Guardians will loose all auras (harmful or not) this breaks many quests, for pets this issue doesn't happen.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/9480287-guardians-evade-issue?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by mik1893**
closes #14347

### CODE REVIEW

**Comment by mik1893** [comment]
Kilyana is testing it now to check the if validity covers all cases

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -4040,6 +4040,23 @@ void Unit::RemoveArenaAuras()
...
 
+void Unit::RemoveAurasOnEvade()
+{
+    for (AuraApplicationMap::iterator iter = m_appliedAuras.begin(); iter != m_appliedAuras.end();)
+    {
+        AuraApplication const* aurApp = iter->second;
+        Aura const* aura = aurApp->GetBase();
+
+        if (!aura->IsPassive() && !aurApp->IsPositive() // all negative, non - passive auras,
+            && !(aura->GetSpellInfo()->AttributesEx3 & SPELL_ATTR3_DEATH_PERSISTENT) && // non death-persistent
```
Comment (DDuarte) [inline src/server/game/Entities/Unit/Unit.cpp]:
```c++
!aura->GetSpellInfo()->HasAttribute(SPELL_ATTR3_DEATH_PERSISTENT)
```

**Comment by Killyana** [comment]
It fix the issue as intended but in some cases like Festergut and Putricide the auras are not removed after a reset.
aura: 
id: 69166 [Inhaled Blight] for Festergut 
id: 71603 [Mutated Strength] and id: 71604 [Mutated Strength] for Putricide
probably many other bosses are affected.

File: src/server/game/Miscellaneous/SharedDefines.h [hunk 2]
```diff
@@ -407,12 +407,12 @@ enum SpellAttr3
...
     SPELL_ATTR3_UNK21                            = 0x00200000, // 21 unused
     SPELL_ATTR3_REQ_WAND                         = 0x00400000, // 22 Req wand
-    SPELL_ATTR3_UNK23                            = 0x00800000, // 23
+    SPELL_ATTR3_UNK23                            = 0x00800000, // 23 
     SPELL_ATTR3_REQ_OFFHAND                      = 0x01000000, // 24 Req offhand weapon
     SPELL_ATTR3_UNK25                            = 0x02000000, // 25 no cause spell pushback ?
     SPELL_ATTR3_CAN_PROC_WITH_TRIGGERED          = 0x04000000, // 26 auras with this attribute can proc from triggered spell casts with SPELL_ATTR3_TRIGGERED_CAN_TRIGGER_PROC_2 (67736 + 52999)
     SPELL_ATTR3_DRAIN_SOUL                       = 0x08000000, // 27 only drain soul has this flag
-    SPELL_ATTR3_UNK28                            = 0x10000000, // 28
+    SPELL_ATTR3_ENRAGE                           = 0x10000000, // 28 flag used for creature enrage abilities (61632,26662 for example)
```
Comment (Shauren) [inline src/server/game/Miscellaneous/SharedDefines.h]: i dont think so. http://paste2.org/I2I2whck

**Comment by mik1893** [comment]
probably a method to get the Enrage spells should be implemented, or a way to call this with a parameter from bosses (@MitchesD add your thoughts)

**Comment by Killyana** [comment]
@mik1893 
Issue with id: 71704 [Create Concoction] (putricide transformation) it's not removed on evade, but removed on death, it must be the opposite.

**Comment by mik1893** [comment]
Attributes: 0x20000100 (SPELL_ATTR0_HIDE_IN_COMBAT_LOG, SPELL_ATTR0_UNAFFECTED_BY_INVULNERABILITY)
AttributesEx1: 0x10000000 (SPELL_ATTR1_DONT_DISPLAY_IN_AURA_BAR)
AttributesEx3: 0x10000000 (SPELL_ATTR3_UNK28)
AttributesEx6: 0x01000000 (SPELL_ATTR6_CAN_TARGET_UNTARGETABLE)

it's should be removed on both cases, but it's an exception. It's difficult to find a common rule for this, the only way i think about is that this behaviour should be applied when it's a non-boss creature, and bosses just gets all auras removed on evade (like it was before basically).

I will try to make a more generic solution for this

**Comment by Killyana** [comment]
According to this video https://youtu.be/IzPe9MeaUU4?t=4m11s the aura persist after death.

**Comment by mik1893** [comment]
ok so this aura is missing the attributed DEATH_PERSISTANT.

**Comment by mik1893** [comment]
for the removal on evade, this is an exception and should be handled in boss spellscript

File: src/server/game/Miscellaneous/SharedDefines.h [hunk 3]
```diff
@@ -407,7 +407,7 @@ enum SpellAttr3
     SPELL_ATTR3_DEATH_PERSISTENT                 = 0x00100000, // 20 Death persistent spells
     SPELL_ATTR3_UNK21                            = 0x00200000, // 21 unused
     SPELL_ATTR3_REQ_WAND                         = 0x00400000, // 22 Req wand
-    SPELL_ATTR3_UNK23                            = 0x00800000, // 23
+    SPELL_ATTR3_UNK23                            = 0x00800000, // 23
```
Comment (Carbenium) [inline src/server/game/Miscellaneous/SharedDefines.h]: Remove the space at the end.

Comment (tkrokli) [inline src/server/game/Miscellaneous/SharedDefines.h]: (which will make this file unneeded in the PR and automatically removed) ;-)

**Comment by ccrs** [comment]
some auras on sindragosa arent removed too (must check for raid mode entries)

**Comment by Rushor** [comment]
someone has to check at least every bossfight and check the remove of the auras on evade

**Comment by ccrs** [comment]
confirm this problem in lanathel too

**Comment by Killyana** [comment]
I don't know if it's possible to do: removing auras that result from spells casted during the combat, this will fix all the issues.

**Comment by mik1893** [comment]
i'm thinking....

**Comment by mik1893** [comment]
I added a property on aura to know if it was added on spawn from creature_addon
test now if it works properly

**Comment by ccrs** [comment]
"removing auras that result from spells casted during the combat"

that should be check, is it ALWAYS that behaviour correct?

**Comment by Killyana** [comment]
Yes when a npc evade he must loose all the auras except the ones set on spawn, of course for pets and guardians all the aura must remain (but still need confirmation)

**Comment by mik1893** [comment]
a pet or guardian never goes in evade, actually, so it wont be a problem

**Comment by Killyana** [comment]
For Pet it's fine, but we need to fix the guardians.

**Comment by mik1893** [comment]
ok, this should be definitive. please test it!

**Comment by Killyana** [comment]
On a recent commit (I cannot find it) Auras from Creature_addon are no longer removed on evade, still in some cases when the aura is set on spawn with SAI, must be fixed on DB  by changing the event on reset or moving the aura to the creature_addon.

But for  the guardians issue it must be fixed as they loose auras on evade, I opened this issue for guardians when the evade issue wasn't fixed for auras from creature_addon so thought it was related.

This commit changes nothing to the actual behavior, sorry for the wasted time.

**Comment by mik1893** [comment]
not needed anymore

**Pull request closed without merging**